alvr/adb/src/lib.rs

```diff
@@ -0,0 +1,393 @@
+#[cfg(target_os = "macos")]
+const PLATFORM_TOOLS_OS: &str = "darwin";
+#[cfg(windows)]
+const PLATFORM_TOOLS_OS: &str = "windows";
+
+const REQUEST_TIMEOUT: Duration = Duration::from_secs(5);
+
+pub enum WiredConnectionStatus {
+    NotReady(String),
+    Ready,
+}
+
+pub fn setup_wired_connection(
+    layout: &Layout,
+    control_port: u16,
+    stream_port: u16,
+    progress_callback: impl Fn(usize, Option<usize>) -> Result<()>,
+) -> Result<WiredConnectionStatus> {
+    let adb_path = require_adb(layout, progress_callback)?;
+
+    let device_serial = match list_devices(&adb_path)?
+        .into_iter()
+        .filter_map(|d| d.serial)
+        .find(|s| !s.starts_with("127.0.0.1"))
+    {
+        None => {
+            return Ok(WiredConnectionStatus::NotReady(
+                "No wired devices found".to_owned(),
+            ))
+        }
+        Some(serial) => serial,
+    };
+
+    let ports = HashSet::from([control_port, stream_port]);
+    let forwarded_ports: HashSet<u16> = list_forwarded_ports(&adb_path, &device_serial)?
+        .into_iter()
+        .map(|f| f.local)
+        .collect();
+    let missing_ports = ports.difference(&forwarded_ports);
+    for port in missing_ports {
+        forward_port(&adb_path, &device_serial, *port)?;
+        dbg_connection!("setup_wired_connection: Forwarded port {port} of device {device_serial}");
+    }
+
+    #[cfg(debug_assertions)]
+    let process_name = "alvr.client.dev";
+    #[cfg(not(debug_assertions))]
+    let process_name = "alvr.client.stable";
+    if get_process_id(&adb_path, &device_serial, process_name)?.is_none() {
+        return Ok(WiredConnectionStatus::NotReady(
+            "ALVR client is not running".to_owned(),
+        ));
+    }
+    if !is_activity_resumed(&adb_path, &device_serial, process_name)? {
+        return Ok(WiredConnectionStatus::NotReady(
+            "ALVR client is paused".to_owned(),
+        ));
+    }
+
+    Ok(WiredConnectionStatus::Ready)
+}
+
+pub fn teardown_wired_connection(layout: &Layout) -> Result<()> {
+    let adb_path = get_adb_path(layout).context("Failed to get ADB executable path")?;
+    kill_server(&adb_path)?;
+    Ok(())
+}
+
+fn get_command(adb_path: &str, args: &[&str]) -> Command {
+    let mut command = Command::new(adb_path);
+    command.args(args);
+
+    #[cfg(windows)]
+    command.creation_flags(0x08000000); // CREATE_NO_WINDOW
+
+    command
+}
+
+fn download(
+    url: &str,
+    progress_callback: impl Fn(usize, Option<usize>) -> Result<()>,
+) -> Result<Vec<u8>> {
+    let agent = ureq::builder()
+        .timeout_connect(REQUEST_TIMEOUT)
+        .timeout_read(REQUEST_TIMEOUT)
+        .build();
+    let response = agent.get(url).call()?;
+    let maybe_expected_size: Option<usize> = response
+        .header("Content-Length")
+        .and_then(|l| l.parse().ok());
+    let mut result = match maybe_expected_size {
+        Some(size) => Vec::with_capacity(size),
+        None => Vec::new(),
+    };
```

Comment:
you can use `.unwrap_or(0)` and feed to `Vec::with_capacity()`, or `.map()` + `.unwrap_or_default()`